### DISCUSSION TITLE
Monitoring idea: Can we support Zipkin or Jaeger?

### DISCUSSION THREAD

**Discussion by tungphuong** [ThreeMammals/Ocelot#1820] [Ideas]
HI,

About tracing feature, can we integrate to zipkin or jaeger? Becase Butterfly is still preview version, we cannot use it for production.

Thanks,

**Comment by TomPallister** [comment]
@tungphuong Ocelot doesnt support zipkin at the moment, you could probably inject middleware or use delegating handlers to add this to Ocelot. I will leave this issue open and look into if we could provide easy integration with zipkin or jaeger.

**Comment by tungphuong** [comment]
Thanks, I am trying to do like that with delegating handlers. Could you please drop me some sample code? I will try

**Reply by raman-m** [reply]
Docs: [Delegating Handlers](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html)

**Comment by tungphuong** [comment]
@TomPallister I know you are ready to support Butterfly. But it is still in preview. So, it is difficult to use for production. So, I suggest we should support Jaeger.

**Reply by raman-m** [reply]
Do we need to support Butterfly in 2023?
Give me the link to active repo or Butterfly team website plz!

**Comment by geffzhang** [comment]
Butterfly should be replace with skywalking-netcore https://github.com/OpenSkywalking/skywalking-netcore， This is docs https://github.com/apache/incubator-skywalking/tree/master/docs

**Reply by raman-m** [reply]
Repo link: **404**
Docs link: redirected to https://github.com/apache/skywalking/tree/master/docs

**Comment by TomPallister** [comment]
OK cool, If anyone wants to pick this work up please feel free to do so!

**Comment by TomPallister** [comment]
started looking at this, I'm going to start breaking the dependencies like butterfly out into there own Ocelot.xxx packages e.g. the first one will be Ocelot.Butterfly. I will have some kind of interface that all the tracing libraries can sustain. This is my general idea and I will see if it works before pushing new versions etc :)

**Comment by TomPallister** [comment]
@geffzhang from looking at Skywalking docs I dont think that you would need to do an Ocelot integration? I will see how it works.

**Comment by geffzhang** [comment]
@TomPallister It was  NET Core Probe。http://www.cnblogs.com/liuhaoyang/p/skywalking-dotnet-v02-release.html
Now we can remove butterfly.

**Reply by raman-m** [reply]
Sorry? 😃

**Comment by tungphuong** [comment]
@TomPallister Do you have plan to support Jeager?

**Reply by raman-m** [reply]
Yes! Welcome for contributions! 😉

**Comment by TomPallister** [comment]
@tungphuong yes in the future, it isn't something I'm working on at the moment. I need to finish #529 before I do this work because I cannot bring more dependencies into Ocelot. Users must be able to pick their dependencies more specifically.

**Comment by tungphuong** [comment]
@TomPallister many thanks, I am waiting for this feature.

**Reply by raman-m** [reply]
We are waiting for your PR... You are welcome!

**Comment by TomPallister** [comment]
No worries, while you are waiting you might be able to implement for your use case using DelegatingHandlers feature in Ocelot but im not 100% sure.

**Comment by kjellegafvelin** [comment]
I'm currently looking in to add OpenTracing support to Ocelot. Is there anyone else doing this too?

**Reply by raman-m** [reply]
Thinking on that... To be fair, [Tracing](https://ocelot.readthedocs.io/en/latest/features/tracing.html) feature is outdated in 2023... 😢

**Comment by tungphuong** [comment]
Currently, I am using "OpenTelemetry" in my project. It is quite good; I will update link then

**Reply by raman-m** [reply]
Sorry? Which link you're talking about?

**Comment by EngRajabi** [comment]
support jaeger and zipkin.
These two pamphlets are the best tools

**Comment by kjellegafvelin** [comment]
With the Open Tracing support I added you can use Jaeger, Zipkin or any other tracing tool as long as it has a .Net client with the Open Tracing interfaces.

@TomPallister maybe it's time to close this one now?

**Reply by raman-m** [reply]
Do you have an intention to contribute to support Jaeger, Zipkin?
And rework [Tracing](https://ocelot.readthedocs.io/en/latest/features/tracing.html) feature with new providers?

**Comment by kaarthickraman** [comment]
@TomPallister should this be closed ?  As per @kjellegafvelin open tracing supports jaeger. Pls update.

**Reply by raman-m** [reply]
Wow!
